core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultShadowVariableSessionFactory.java

```diff
@@ -207,16 +217,22 @@ public List<VariableUpdaterInfo<Solution_>> getUpdatersForEntity(Object entity)
 
     private static <Solution_> Map<VariableMetaModel<Solution_, ?, ?>, GroupVariableUpdaterInfo<Solution_>>
             getGroupVariableUpdaterInfoMap(
-                    List<DeclarativeShadowVariableDescriptor<Solution_>> declarativeShadowVariableDescriptors) {
+                    List<DeclarativeShadowVariableDescriptor<Solution_>> declarativeShadowVariableDescriptors,
+                    Object[] entities) {
         var sortedDeclarativeVariableDescriptors =
                 topologicallySortedDeclarativeShadowVariables(declarativeShadowVariableDescriptors);
         var groupIndexToVariables = new HashMap<Integer, List<DeclarativeShadowVariableDescriptor<Solution_>>>();
         var groupVariables = new ArrayList<DeclarativeShadowVariableDescriptor<Solution_>>();
         groupIndexToVariables.put(0, groupVariables);
         for (var declarativeShadowVariableDescriptor : sortedDeclarativeVariableDescriptors) {
-            if (!groupVariables.isEmpty() && Arrays.stream(declarativeShadowVariableDescriptor.getSources())
-                    .anyMatch(rootVariableSource -> rootVariableSource.parentVariableType() == ParentVariableType.GROUP)) {
-                // Create a new variable group, since the group might reference prior variables
+            if (!groupVariables.isEmpty() && (Arrays.stream(declarativeShadowVariableDescriptor.getSources())
```

Comment:
This condition is becoming kinda crazy. Let's split it into its component parts, name the variables, that way we can better read what it does.